tests/test_queryselector.py

```diff
@@ -0,0 +1,301 @@
+    page = isolated_page
+    await page.setContent('<section>hello</section><div> world</div>')
+    divHandle = await page.J('div')
+    text = await page.Jeval('section', "(e, div) => e.textContent + div.textContent", divHandle)
+    assert text == 'hello world'
+
+
+@chrome_only
+@sync
+async def test_Jeval_throws_exc_if_no_element(isolated_page):
+    """Test Page().Jeval() should throw error if no element is found."""
+    page = isolated_page
+    with pytest.raises(ElementHandleError) as exc:
+        await page.Jeval('section', "e => e.id")
+    assert 'Error: failed to find element matching selector "section"' in str(exc)
+
+
+@chrome_only
+@sync
+async def test_page_JJeval_executes_js_func(isolated_page):
+    """Test Page().JJeval() method executes JS function
+    with all elements which matches ``selector``.
+
+    """
+    page = isolated_page
+    await page.setContent('<div>hello</div><div>beautiful</div><div>world!</div>')
+    divsCount_jjeval = await page.JJeval('div', "divs => divs.length")
+    divsCount_query = await page.querySelectorAllEval('div', "divs => divs.length")
+    assert divsCount_jjeval == divsCount_query == 3
+
+
+@chrome_only
+@sync
+async def test_page_J_queries_element(isolated_page):
+    """Test Page().J() method should query existing element."""
+    page = isolated_page
+    await page.setContent('<section>test</section>')
+    element = await page.J('section')
+    assert element
+
+
+@chrome_only
+@sync
+async def test_J_returns_none_if_no_element(isolated_page):
+    """Test Page().J() method should return null for non-existing element."""
+    page = isolated_page
+    element = await page.J('non-existing-element')
+    assert not element
+
+
+@chrome_only
+@sync
+async def test_page_JJ_queries_elements(isolated_page):
+    """Test Page().JJ() method should query existing elements."""
+    page = isolated_page
+    await page.setContent('<div>A</div><br/><div>B</div>')
+    elements = await page.JJ('div')
+    assert len(elements) == 2
+    assert [await page.evaluate("e => e.textContent", element) for element in elements] == ['A', 'B']
+
+
+@chrome_only
+@sync
+async def test_page_JJ_returns_empty_list_if_no_element(isolated_page, server):
+    """Test Page().JJ() method should return empty list if nothing is found."""
+    page = isolated_page
+    await page.goto(server / "empty.html")
+    elements = await page.JJ('div')
+    assert not elements and elements == []
+
+
+@chrome_only
+@sync
+async def test_Jx_queries_elements(isolated_page):
+    """Test Page().Jx() method should query existing elements."""
+    page = isolated_page
+    await page.setContent('<section>test</section>')
+    elements_jx = await page.Jx('/html/body/section')
+    elements_path = await page.xpath('/html/body/section')
+    assert isinstance(elements_jx, list)
+    assert isinstance(elements_path, list)
+    assert elements_jx[0]
+    assert elements_path[0]
+    assert len(elements_path) == len(elements_jx) == 1
```

Comment:
Since these (should) refer to the exact same function, this could be simplified a bit:
```suggestion
    assert elements_jx == elements_path
    assert isinstance(elements_jx , list)
    assert elements_jx[0]
    assert len(elements_jx) == 1
```
ehh, looking at it again not really sure which is better. Your call on this one.

Reply:
I've checked `    assert elements_jx == elements_path` and it's false, as they are two different objects, therefore two different reference.
I agree my solution is a little bit ugly, but have no better idea.

Reply:
That's strange, if they are lists then the equality of them shouldn't be different, unless the list is filled with `ElementHandle`s or something? Regardless, not really a big deal in this case.